File: Java-PerfKata/src/test/java/perfkata/parrot/a20171207/Parrot.java
package perfkata.parrot.a20171207;

public class Parrot {

    private ParrotTypeEnum type;
    private double voltage;
    private boolean isNailed;

    public Parrot(ParrotTypeEnum type, int numberOfCoconuts, double voltage, boolean isNailed) {
        this.type = type;
        this.voltage = voltage;
        this.isNailed = isNailed;
    }

    public double getSpeed() {
        switch (type) {
            case NORWEGIAN_BLUE:
                return (isNailed) ? 0 : getBaseSpeed(voltage);
            default:
                throw new RuntimeException("Should be unreachable");
        }
    }

    private double getBaseSpeed(double voltage) {
        return Math.min(24.0, voltage * getBaseSpeed());
    }

    protected double getBaseSpeed() {
        return 12.0;
    }

}

File: JavaScript/src/parrot.js
export const PARROT_TYPES = {
    NORWEGIAN_BLUE: 'NORWEGIAN_BLUE',
};

export class Parrot {
    constructor(type, numberOfCoconuts, voltage, isNailed) {
    }

    getSpeed() {
        throw new Error("Should be unreachable");
    }

    getBaseSpeed() {
        return 12;
    }
}

export class EuropeanParrot extends Parrot {
    constructor() {
        super(null, 0, 0, false);
    }

    getSpeed() {
        return this.getBaseSpeed();
    }
}

export class AfricanParrot extends Parrot {
    constructor(numberOfCoconuts) {
        super(null, 0, 0, false);
        this.numberOfCoconuts = numberOfCoconuts;
    }

    getSpeed() {
        return Math.max(0, this.getBaseSpeed() - this.getLoadFactor() * this.numberOfCoconuts);
    }

    getLoadFactor() {
        return 9;
    }
}

export class NorwegianBlueParrot extends Parrot {
    constructor(type, numberOfCoconuts, voltage, isNailed) {
        super(null, 0, 0, false);
        this.voltage = voltage;
        this.isNailed = isNailed;
    }

    getSpeed() {
        return this.isNailed ? 0 : this.speedWithVoltage(this.voltage);
    }

    speedWithVoltage(voltage) {
        return Math.min(24, voltage * this.getBaseSpeed());
    }
}
